Improvements to OpenAPI aggregation

mbohlool opened the pull request:
Fixes #50863
Fixes #50011
Related: #50896

mbohlool committed:
Consolidate local OpenAPI specs and APIServices' spec into one data structure

Remove APIService OpenAPI spec when it is deleted

Add eTag support and returning httpStatus to OpenAPI spec downloader

Update aggregated OpenAPI spec periodically

Use delegate chain

Refactor OpenAPI aggregator to have separate controller and aggregation function

Enable OpenAPI spec for extensions api server

Do not filter paths. higher priority specs wins the conflicting paths

Move OpenAPI aggregation controller to pkg/controller/openapi

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -110,6 +138,10 @@ type byPriority struct {
 func (a byPriority) Swap(i, j int) { a.specs[i], a.specs[j] = a.specs[j], a.specs[i] }
 func (a byPriority) Less(i, j int) bool {
+	// All local specs will come first
+	if a.specs[i].apiService.Spec.Service == nil {
```
is this how we operate now in the aggegator? So the priorities in https://github.com/sttts/kubernetes/blob/43b8715d82a713bb2bff5dc53d5e213a42d9d115/cmd/kube-apiserver/app/aggregator.go#L170 do not really play role in comparison to external api services?

mbohlool replied:
Well, both yes and no. I believe those priorities is copied from what `kubeApiServer` already has. Problem is we do not register those separately with aggregator. We get spec from `kubeApiServer` that has all of those. We can change implementation a little to take those into account but I think it is out of this PR (This PR does not actually change any priority from what it was before).

mbohlool replied:
Having said that, it is totally possible to take those priorities into account. Maybe we should create a separate issue for it and see if we can get it done in 1.8.

sttts replied:
Building something inconsistent here doesn't feel good. IMO we have to follow the architecture of the API aggregation much more than working around it and creating an approximation of the logic on the API side. Not that I have the perfect solution here though either :-/

mbohlool replied:
I completely agree. What I tried to say is: this is outside of the scope of this PR (The problem was not introduced here as far as I can tell. The line added here is merely simulate what we were doing before this PR that aggregating `kubeApiServer` as whole with aggregator server's spec then proceed to any non-local spec). We should definitely find a solution for this and fix it.

mbohlool commented:
@sttts There were some comments about the ordering of local API Services and if we need to take that into account for OpenAPI aggregation. I want to step back and look at the whole ordering merge. Why do we need it? We needed it because we were worried about definition renaming and we wanted to higher priority API Services has more consistent spec. But it is very unlikely for local specs to have model conflicts. I can even claim that it is a bug if we have such a conflict. We do use full names for model names and conflicts can only happens if we use different version of types in different API services, that should not happen for local specs as they all should be compiled from one version. I propose we disable model renaming for local API services unless there is a use-case for conflicts of models for local specs. WDYT?

sttts commented on staging/src/k8s.io/apiserver/pkg/server/genericapiserver.go:
```diff
@@ -165,6 +168,9 @@ type DelegationTarget interface {
 	ListedPaths() []string
+
+	// NextDelegate returns the next delegationTarget in the chain of delegations
+	NextDelegate() DelegationTarget
```
:+1:

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -47,36 +50,65 @@ type openAPIAggregator struct {
 
-	contextMapper request.RequestContextMapper
+func (s *openAPIAggregator) addLocalSpec(spec *spec.Swagger, localHandler http.Handler, name, etag string) {
+	localApiService := apiregistration.APIService{}
```
nit: `Name: name`

mbohlool replied:
The same comment as before. It would look like this with a new import if I move Name: name inside:

```golang
import metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
...
	localApiService := apiregistration.APIService{
		ObjectMeta: metav1.ObjectMeta {
			Name: name,
		},
	}
```
I prefer two existing to line of code with no extra import.

sttts replied:
sorry, missed the previous comment

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -47,36 +50,65 @@ type openAPIAggregator struct {
 		openAPISpecs:  map[string]*openAPISpecInfo{},
 		contextMapper: contextMapper,
 	}
+	s.openAPIAggregationController = NewOpenAPIAggregationController(s)
```
nit: move up into the struct

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -47,36 +50,65 @@ type openAPIAggregator struct {
+		// First spec is the base for all other specs. Keep all of its path
+		// e.g. /api/ from this spec. For any other spec after first one we
+		// should only be interested in things in /apis/ path.
+		if i != 0 {
```
the /api assumption is not generally true for `i == 0`. Can we look for the first delegate which serves `/api` instead? All after that get filtered.

mbohlool replied:
Fixed with next commit.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -47,36 +50,65 @@ type openAPIAggregator struct {
+	aggregator.FilterSpecByPaths(aggregatorOpenAPISpec, []string{"/apis/"})
+
+	// Reserving non-name spec for aggregator's Spec.
+	s.addLocalSpec(aggregatorOpenAPISpec, nil, fmt.Sprintf(localDelegateChainNamePattern, i), "")
```
using the `i` here is strange. I still think it's much more natural to add this spec first (using `i == 0`) and then all the others. That's the natural order.

mbohlool replied:
Fixed with next commit.

sttts replied:
Which is? Hard to see here in Github's review interface.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -110,6 +144,11 @@ type byPriority struct {
 func (a byPriority) Less(i, j int) bool {
+	// All local specs will come first
+	// WARNING: This will result in not following priorities for local APIServices.
+	if a.specs[i].apiService.Spec.Service == nil {
```
this means that the order of local services is not stable. This means that not only the APIService priorities are ignored, but also the spec might change over time. We should at least use the delegate order as priority and sort here accordingly. Then we get a stable order which corresponds with the delegation order.

mbohlool replied:
I've commented on this PR about this issue. I think order should not matter for local delegates at all. I sent this PR (https://github.com/kubernetes/kube-openapi/pull/4) to make it possible to Merge specs without renaming definitions. All local delegates should use newly introduced SafeMergeSpec and if we do so, we don't need to worry about order (I would even move them out of this sorting if we agree on this).

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -141,16 +183,21 @@ func sortByPriority(specs []openAPISpecInfo) {
+	if !exists {
+		return nil, fmt.Errorf("localDelegate spec is missing")
+	}
+	specToReturn, err = aggregator.CloneSpec(localDelegateSpec.spec)
```
see my other comment about using the delegation position as priority for the local services. Then we can get rid of the special handling here.

mbohlool replied:
ditto

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -217,28 +269,48 @@ func (s *openAPIAggregator) handlerWithUser(handler http.Handler, info user.Info
 	switch writer.respCode {
+	case http.StatusNotModified:
+		if len(etag) == 0 {
+			return nil, etag, http.StatusNotModified, fmt.Errorf("http.StatusNotModified is not allowed in absense of etag")
```
absence

sttts replied:
Does this mean a server with an unexpected `StatusNotModified` will kill our openapi aggregation?

mbohlool replied:
Mostly No. If it consistently has this behaviour, the APIServer will not be registered. If it start behaving like that after first attempt, the aggregation won't update from that point on until that API server is disabled.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -217,28 +269,48 @@ func (s *openAPIAggregator) handlerWithUser(handler http.Handler, info user.Info
+		}
+		etag = writer.Header().Get("Etag")
+		if len(etag) == 0 {
+			etag = etagFor(writer.data)
```
I wouldn't guess an etag here. We are the client and should not guess the etag mechanism of the server.

mbohlool replied:
We are not guessing etags, we are supporting our own. When server does not return one, we compute our own so we can skip expensive merging if the spec is not changed. It (should, a bug there, I will fix it) acts as a change detection with the same way as if server supports etag but we will always download the spec.

sttts replied:
Makes sense, please add a comment that we do that for that reason of internal comparison. This also means the etag is passed to the server on next download. Feels like a leak. Shouldn't harm though I agree.

mbohlool replied:
I sent the fix for this that fully support and differentiate local and remote etags.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -217,28 +269,48 @@ func (s *openAPIAggregator) handlerWithUser(handler http.Handler, info user.Info
+		return openApiSpec, etag, http.StatusOK, nil
 	default:
-		return nil, fmt.Errorf("failed to retrive openAPI spec, http error: %s", writer.String())
+		return nil, "", 0, fmt.Errorf("failed to retrive openAPI spec, http error: %s", writer.String())
```
same question as above: will a server with a wrong http code kill our aggregation? Shouldn't we be gracefuller?

sttts replied:
Shouldn't we return the right http code even here? not `0`.

mbohlool replied:
I've commented on the function to say the httpCode is only valid if err == nil. We don't want to confuse the caller by providing both err and httpCode giving them the impression that httpCode can be reliable even if err is not nil. The caller should always check for err first.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -250,21 +322,74 @@ func (s *openAPIAggregator) loadApiServiceSpec(handler http.Handler, apiService
 	}
+	if openApiSpec == nil {
+		// API service does not provide an OpenAPI spec, ignore it.
+		return nil
```
shouldn't we delete anything from `s.openAPISpecs`?

mbohlool replied:
Fixed with next commit.

mbohlool replied:
fixed.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_controller.go:
```diff
@@ -0,0 +1,90 @@
+	UpdateApiServiceSpec(apiServiceName string) (changed, deleted bool, err error)
+}
+
+type OpenAPIAggregationController struct {
```
godoc please. What does this controller do?

mbohlool replied:
Fixed.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_controller.go:
```diff
@@ -0,0 +1,90 @@
+	"k8s.io/client-go/util/workqueue"
+)
+
+type OpenAPIAggregationManager interface {
```
godoc

mbohlool replied:
Fixed.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_controller.go:
```diff
@@ -0,0 +1,90 @@
+	openAPIAggregationManager OpenAPIAggregationManager
+	queue                     workqueue.RateLimitingInterface
+}
+
```
does this struct implement the interface? If yes, add a type assertion here.

mbohlool replied:
By this struct you mean the controller? no it does not. `openAPIAggregator` does and I added a type assertion there.

sttts replied:
👍

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -250,21 +322,74 @@ func (s *openAPIAggregator) loadApiServiceSpec(handler http.Handler, apiService
+	return nil
+}
+
+func (s *openAPIAggregator) UpdateApiServiceSpec(apiServiceName string) (shouldBeRateLimited, deleted bool, err error) {
```
this func is not thread save. This should be at least documented, especially as it called from the controller.

mbohlool replied:
I made it thread safe in next commit.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_controller.go:
```diff
@@ -0,0 +1,90 @@
+		shouldBeRateLimited, deleted, err := c.openAPIAggregationManager.UpdateApiServiceSpec(key.(string))
+
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("%v failed with : %v", key, err))
```
continue. Otherwise, you run into the `!deleted` branch below.

mbohlool replied:
It is intentional. If err is not nil we only log it here, but the flags returned from Update decide what to do. In case of most of errors, `shouldBeRateLimited` has been set to check again if the error fixed but rate limit it. There could be cases of errors that we don't want to retry for (we would set `deleted` flag for them, I would rename this to `notRetry`) or maybe we do not want to rate limit them. Right now the update method returns rate limit flag for all errors.

sttts replied:
Got it.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -250,21 +322,74 @@ func (s *openAPIAggregator) loadApiServiceSpec(handler http.Handler, apiService
 		delete(s.openAPISpecs, apiService.Name)
 		return err
 	}
+	if !existingService {
```
why this re-enqueue?

mbohlool replied:
What do you mean by "re-"? We only enqueue if the service does not exists.

sttts replied:
My understanding is/was that the controller downloads the specs. And above we already did that. So why should be tell the controller to do it again?

mbohlool replied:
we only add enqueue it to the controller if it does not exists. This will happened only first time we Load spec for APIService.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -250,21 +322,74 @@ func (s *openAPIAggregator) loadApiServiceSpec(handler http.Handler, apiService
 }
+
+func (s *openAPIAggregator) removeApiServiceSpec(apiServiceName string) error {
+	if _, existingService := s.openAPISpecs[apiServiceName]; existingService {
```
return early

mbohlool replied:
fixed.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:
```diff
@@ -307,6 +313,7 @@ func (s *APIAggregator) RemoveAPIService(apiServiceName string) {
 	}
 	s.GenericAPIServer.Handler.NonGoRestfulMux.Unregister(proxyPath)
 	s.GenericAPIServer.Handler.NonGoRestfulMux.Unregister(proxyPath + "/")
+	s.openAPIAggregator.removeApiServiceSpec(apiServiceName)
```
this is called from another thread/go-routine than the new controller. Both write to `s.openAPISpecs`, without locking 🙈

mbohlool replied:
Fixed. All public methods of openapiAggregator are now thread safe and I changed this to call the public method.

sttts replied:
missing a commit? Don't see the locking yet.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -250,21 +322,74 @@ func (s *openAPIAggregator) loadApiServiceSpec(handler http.Handler, apiService
 		return nil
```
Also here we are completely unprotected by locks, but called from several go-routines.

sttts replied:
Also we are called her by the aggregator synchronously. If this download blocks, the main aggregation controller blocks. Do we want that? /cc @deads2k

mbohlool replied:
All public methods are lock protected now.

deads2k replied:
> Also we are called her by the aggregator synchronously. If this download blocks, the main aggregation controller blocks. Do we want that? /cc @deads2k

We definitely do not.  This should be a separate controller, unreleated to the main api aggregation controller.

mbohlool replied:
Fixed. Now only OpenAPI controller loads/updates specs.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -250,21 +322,74 @@ func (s *openAPIAggregator) loadApiServiceSpec(handler http.Handler, apiService
+	if err != nil {
+		return true, false, err
+	}
+	if httpStatus == http.StatusNotModified {
```
a `switch` statement is easier to read here

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -217,28 +269,48 @@ func (s *openAPIAggregator) handlerWithUser(handler http.Handler, info user.Info
+		}
+		return nil, etag, http.StatusNotModified, nil
+	case http.StatusNotFound:
+		// Gracefully skip 404, assuming the server won't provide any spec
```
the logic in the callers is more like: "assuming the server will return one later, let's retry".

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -250,21 +322,74 @@ func (s *openAPIAggregator) loadApiServiceSpec(handler http.Handler, apiService
+		return false, false, nil
+	}
+	if httpStatus == http.StatusNotFound || spec == nil {
+		return true, false, nil
```
this means to retry?

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -250,21 +322,74 @@ func (s *openAPIAggregator) loadApiServiceSpec(handler http.Handler, apiService
+	oldSpec := specInfo.spec
+	specInfo.spec = spec
+	if err := s.updateOpenAPISpec(); err != nil {
+		specInfo.spec = oldSpec
```
why this acrobatic with `oldSpec`. Cannot follow....

mbohlool replied:
I tried to keep the logic that if a function fails, it won't change the state of anything. Here I keep old spec and restore it if we fail to update new one.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_controller.go:
```diff
@@ -0,0 +1,90 @@
+	keys := []string{}
+	defer func() {
+		for _, key := range keys {
+			c.queue.Done(key)
```
shouldn't we call `Done` in each inner loop? All other controllers I looked at do that. Note: `queue.Get` blocks.

mbohlool replied:
Not calling done was intentional. Otherwise we keep looping as we add items back into the queue.

sttts replied:
cannot follow. The `Get()` below blocks forever. We only stop the loop on shutdown or panic. Before that elements are never actually re-added to the queue (i.e. on APIService updates). Why is looping a problem in a worker? Isn't that the task of a work?

mbohlool replied:
ah, you are right, I miss-read the Get code. Will fix this.

frobware commented:
@mbohlool - just a heads-up - I don't believe the `mergeSpecs()` is working properly. If you run the tests in:

    k8s.io/kubernetes/cmd/kube-apiserver/app/testing

you'll see the following error message:

    Failed to create server chain: Model name conflict in merging OpenAPI spec: io.k8s.kube-aggregator.pkg.apis.apiregistration.v1beta1.APIService

assuming you cherry-pick #51085.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:
```diff
@@ -145,6 +145,8 @@ func (c *Config) SkipComplete() completedConfig {
 }
 
 // New returns a new instance of APIAggregator from the given config.
+// delegationChain is the list of all delegations. The first one in the list will be used to delegate all unknown traffic
```
comment is not updated I guess

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -217,54 +273,143 @@ func (s *openAPIAggregator) handlerWithUser(handler http.Handler, info user.Info
+		return openApiSpec, etag, http.StatusOK, nil
 	default:
-		return nil, fmt.Errorf("failed to retrive openAPI spec, http error: %s", writer.String())
+		return nil, "", 0, fmt.Errorf("failed to retrive openAPI spec, http error: %s", writer.String())
```
retrieve

mbohlool commented:
@sttts Sorry, I forgot some commits. I think I addressed all of your comment even if I didn't say it explicitly on each of them. Please take another look.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -217,54 +297,157 @@ func (s *openAPIAggregator) handlerWithUser(handler http.Handler, info user.Info
+	oldSpec := specInfo.spec
+	specInfo.spec = spec
+	if err := s.updateOpenAPISpec(); err != nil {
+		specInfo.spec = oldSpec
```
regarding your answer to my comment (thanks Github it's hidden now): does `updateOpenAPISpec` change specInfo? If it does, this code is hard to follow, not good. Can we improve that? Why does updateOpenAPISpec change anything on error? The ownership of an object in `s.openAPISpec` should be clear. If possible only atomic changes to the map should be done, not inside object which might be expected here and every mutation behind the scene is surprising.

mbohlool replied:
No, `updateOpenAPISpec` does not change specInfo. What we are doing here is we set the spec of API Service to the new one, try to aggregate it, if we failed to aggregated (possible because of merge problems or deserialization problems, etc.) that means we still serving the old spec for this APIService as we didn't change aggregated spec (that's the logic of don't change anything on error for `updateOpenAPISpec`). So we need to restore old spec in specInfo to have the right input for aggregation. If we don't do this, possibly, all later aggregations will fail as we have a bad input in specInfo for this APIService. Does that make sense?

sttts replied:
So we modify the global state here, not in updateOpenAPISpec. Please add a comment that we are actually changing the state stored in `s.openAPISpecs`. And that we restore it on error.

mbohlool replied:
Done.

mbohlool commented:
This is not a WIP anymore and should be ready to merge. Biggest thing I addressed in to the comments is to remove LoadAPIServiceSpec in OpenAPI aggregator and add a AddAPIServiceSpec method that adds/enqueues the spec into OpenAPI controller to be loaded.

mbohlool commented:
/test pull-kubernetes-verify

test timed out.

mbohlool commented:
/test pull-kubernetes-e2e-gce-etcd3

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -217,54 +300,144 @@ func (s *openAPIAggregator) handlerWithUser(handler http.Handler, info user.Info
+		handler:    handler,
+		etag:       "",
 	}
+	s.openAPIAggregationController.AddAPIService(apiService.Name)
```
much better to delegate here to the controller :+1:

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -217,54 +300,144 @@ func (s *openAPIAggregator) handlerWithUser(handler http.Handler, info user.Info
+	oldSpecs := s.openAPISpecs
+	delete(s.openAPISpecs, apiServiceName)
+	err := s.updateOpenAPISpec()
 	if err != nil {
```
can this really happen? Is it because there might be dangling references to definitions of the deleted spec?

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -217,54 +300,144 @@ func (s *openAPIAggregator) handlerWithUser(handler http.Handler, info user.Info
+		specInfo.spec = oldSpec
+		return true, false, err
+	}
+	glog.Infof("OpenAPI spec loaded for APIService %s.", apiServiceName)
```
less log level would be appropriate I think, maybe 4?

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_controller.go:
```diff
@@ -0,0 +1,111 @@
+/*
+Copyright 2016 The Kubernetes Authors.
```
2017

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -217,54 +300,144 @@ func (s *openAPIAggregator) handlerWithUser(handler http.Handler, info user.Info
+
+// UpdateApiServiceSpec updates and aggregation OpenAPI spec for given service name. It is thread safe.
+func (s *openAPIAggregator) UpdateApiServiceSpec(apiServiceName string) (shouldBeRateLimited, deleted bool, err error) {
+	s.mutex.Lock()
```
this lock is hold during the whole download (with 60s timeout)? So Add/RemoveApiServiceSpec are still blocked by the download.

sttts replied:
Also this means that we cannot do parallel downloads.

sttts commented:
@mbohlool it's already much better. But I believe there is still a serious issue with the lock and the download. Also consider to move the whole downloading logic into the controller. The aggregator actually shouldn't care about downloads, it should aggregate IMO.

mbohlool commented:
@frobware I think the problem was a wrong file checked in into vendor folder (for kube-openapi). I think it fixed. Can you confirm?

frobware commented:
> @frobware I think the problem was a wrong file checked in into vendor folder (for kube-openapi). I think it fixed. Can you confirm?

Yes this aspect is fixed.

However, if you were to `GET /swagger.json` now that we have the delegation chain in place I do not see any entries for the apiserver extensions. Specifically, I do not see any entry for `apiextensions.k8s.io`. I was expecting these to show up now that we enumerate via `NextDelegate()`.

I was looking into this today and the apiextensions server uses `genericapiserver.EmptyDelegate` which returns `nil` for its HTTP handler so no spec can be downloaded in `buildAndRegisterOpenAPIAggregator()` as we `continue` over handlers that are `nil`.

This PR should really have a unit test to assert this.

mbohlool commented:
@frobware  the `EmptyDelegate` is the next in chain after apiextensions server. so if apiextensions server provide an spec (that should provide the spec you are looking for) it should be aggregated. Most probably apiextensions server does not expose OpenAPI spec. Will look into it. anyway, most probably, it shouldn't be related to this code.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -132,25 +192,32 @@ func sortByPriority(specs []openAPISpecInfo) {
-	specToReturn, err = aggregator.CloneSpec(s.inProcessDelegatesOpenAPISpec)
+	localDelegateSpec, exists := s.openAPISpecs[s.baseForMergeSpecName]
+	if !exists {
+		return nil, fmt.Errorf("Cannot find base spec for merging.")
```
nit: lower-case

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -36,47 +36,98 @@ import (
+		// base for the merge. Any other spec would be filtered for their "/apis/"
+		// endpoints only.
+		filterPaths := true
+		if len(s.baseForMergeSpecName) == 0 {
```
do we check that we eventually find one? On the other hand, having not `/api` is a completely normal case. Then we probably should choose the first delegate. Or maybe we should just mark this one (with a bool in the spec info that this provides the /api? Then we can filter accordingly.

sttts replied:
Not even sure that we do need this bool. If we filter out the `/api` paths from all other specs than the first having `/api`, we are fine, aren't we? Then order in `buildOpenAPISpec` for the local services does not matter anymore. Wdyt?

sttts replied:
Or can this even happen that two delegates have `/api`? It shouldn't. If it does it's a conflict that should be fatal for the aggregation.

mbohlool replied:
Hmm, if the case with no `/api` is a valid case, then we can simply point `baseForMergeSpecName` to first delegate. The question is should it be aggregation server or the one after? In normal kubernetes case, is the one after. I thought about adding a boolean, but then I need to make sure one of APIServers will have that boolean. It is just scream bug :) having `baseForMergeSpecName` will make sure we would not have two server having the boolean set. I think we can proceed with setting aggregation API server as the default one if no one provides `/api`. We do get other non-api information from this server (like Swagger.Info, ...) so we need to have one.

sttts replied:
Does anything speak against using the aggregator as base all the time? Does it have all "non-api information" ?

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -173,98 +240,81 @@ func (s *openAPIAggregator) updateOpenAPISpec() error {
-
-func (r *inMemoryResponseWriter) WriteHeader(code int) {
-	r.respCode = code
+// tryUpdatingServiceSpecs tries updating openAPISpecs map, and keep the map intact if the update fails.
```
nit: keep**s**

mbohlool replied:
fixed in next commit.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -173,98 +240,81 @@ func (s *openAPIAggregator) updateOpenAPISpec() error {
+func (s *openAPIAggregator) tryUpdatingServiceSpecs(updateFunc func() error) error {
+	// Keep a backup value and restore it if aggregation failed.
+	backupMap := s.openAPISpecs
+	if err := updateFunc(); err != nil {
```
👍  here in a separate func

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -173,98 +240,81 @@ func (s *openAPIAggregator) updateOpenAPISpec() error {
+// note that map's elements are pointers, so the update method should not change them.
+func (s *openAPIAggregator) tryUpdatingServiceSpecs(updateFunc func() error) error {
+	// Keep a backup value and restore it if aggregation failed.
+	backupMap := s.openAPISpecs
```
This does not work: you only copy the pointer to the map. You have to iterate through it and copy each key-value pair.

mbohlool replied:
what about `backupMap := *s.openAPISpecs`. The thing is updateFunc should not mutate items in the map, it can only creates a new one or remove one.

sttts replied:
Propose that for Go 2.0. I like it :)

mbohlool replied:
fixed in next commit.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -173,98 +240,81 @@ func (s *openAPIAggregator) updateOpenAPISpec() error {
-		s += fmt.Sprintf(", Body: %s", string(r.data))
+	specInfo, existingService := s.openAPISpecs[apiServiceName]
+	if !existingService {
+		return fmt.Errorf("APIService %s does not exists", apiServiceName)
```
nit: %q

mbohlool replied:
fixed in next commit.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -173,98 +240,81 @@ func (s *openAPIAggregator) updateOpenAPISpec() error {
 
-// loadApiServiceSpec loads OpenAPI spec for the given API Service and then updates aggregator's spec.
-func (s *openAPIAggregator) loadApiServiceSpec(handler http.Handler, apiService *apiregistration.APIService) error {
+// RemoveApiServiceSpec removes an api service from OpenAPI aggregation. It is thread safe.
```
add: if it does not exist, no error is returned.

mbohlool replied:
fixed in next commit.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -173,98 +240,81 @@ func (s *openAPIAggregator) updateOpenAPISpec() error {
-	}
+// GetApiServiceSpec returns api service spec info
+func (s *openAPIAggregator) GetApiServiceInfo(apiServiceName string) (handler http.Handler, etag string, exists bool) {
+	s.mutex.Lock()
```
Would a RWMutex make sense here?

mbohlool replied:
fixed in next commit.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_controller.go:
```diff
@@ -0,0 +1,150 @@
+	}
+	returnSpec, newEtag, httpStatus, err := c.downloader.Download(handler, etag)
+	rateLimit := false
+	switch {
```
much nicer to have the download error handling here 👍

sttts commented on cmd/kube-apiserver/app/server.go:
```diff
@@ -165,6 +165,9 @@ func CreateServerChain(runOptions *options.ServerRunOptions, stopCh <-chan struc
 	// this wires up openapi
 	kubeAPIServer.GenericAPIServer.PrepareRun()
 
+	// This will wire up openapi for extension api server too.
```
s/ too//

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -36,47 +35,72 @@ import (
 )
 
 type openAPIAggregator struct {
+	// mutex protects All members of this struct.
```
nit: all

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -173,98 +217,98 @@ func (s *openAPIAggregator) updateOpenAPISpec() error {
+// if the update fails.
+func (s *openAPIAggregator) tryUpdatingServiceSpecs(specInfo *openAPISpecInfo) error {
+	orgSpecInfo, exists := s.openAPISpecs[specInfo.apiService.Name]
+	s.openAPISpecs[specInfo.apiService.Name] = specInfo
```
why this line?

mbohlool replied:
we are trying specInfo (the input of the method, I no longer accept an updateFunc).

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -132,36 +166,46 @@ func sortByPriority(specs []openAPISpecInfo) {
+		if specInfo.spec == nil {
+			continue
+		}
+		specs = append(specs, openAPISpecInfo{specInfo.apiService, specInfo.spec, specInfo.handler, specInfo.etag})
```
specs = append(specs, *spec) ?

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -132,36 +166,46 @@ func sortByPriority(specs []openAPISpecInfo) {
+	if len(specs) == 0 {
+		return &spec.Swagger{}, nil
 	}
 	sortByPriority(specs)
```
we need the order of the local specs now, ordered by position in the chain.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -110,6 +138,12 @@ type byPriority struct {
+	// WARNING: This will result in not following priorities for local APIServices.
+	if a.specs[i].apiService.Spec.Service == nil {
+		// Sort local specs with their name
+		return a.specs[i].apiService.Name < a.specs[j].apiService.Name
```
we have to sort by position in the chain. Aggregator first.

sttts replied:
we do. can you comment that this is the order in the chain?

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -132,36 +166,46 @@ func sortByPriority(specs []openAPISpecInfo) {
 	sortByPriority(specs)
 	for _, specInfo := range specs {
-		if err := aggregator.MergeSpecs(specToReturn, specInfo.spec); err != nil {
+		// TODO: Make kube-openapi.MergeSpec(s) accept nil or empty spec as destination and just clone the spec in that case.
```
indeed, would be slick.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -110,6 +138,12 @@ type byPriority struct {
 func (a byPriority) Swap(i, j int) { a.specs[i], a.specs[j] = a.specs[j], a.specs[i] }
 func (a byPriority) Less(i, j int) bool {
+	// All local specs will come first
+	// WARNING: This will result in not following priorities for local APIServices.
```
If we want this, this will be part of a follow-up. For now this looks good enough.

sttts commented:
Looks much much better now. I like the new structure with the downloader. Thanks for the patience and fixing everything I commented on!

Can you squash?

@deads2k can you look over it and approve?

deads2k commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go:
```diff
@@ -243,7 +248,8 @@ func (s *APIAggregator) AddAPIService(apiService *apiregistration.APIService) er
 	if proxyHandler, exists := s.proxyHandlers[apiService.Name]; exists {
 		proxyHandler.updateAPIService(apiService)
-		return s.openAPIAggregator.loadApiServiceSpec(proxyHandler, apiService)
+		s.openAPIAggregationController.UpdateAPIService(proxyHandler, apiService)
```
reading top to bottom, this surprised me since I would have expected the openapiaggregation controller to drive based on its own sharedinformer and only aggregate things that are `Available` and un-aggregate them when they are not available.  That would avoid coupling the bits, right?

sttts replied:
Good idea :+1:

mbohlool replied:
How about this?
1- OpenAPISpecStore (implements store functionality)
2- Controller keeps the Queue and Key structure it has, and Add/Update/Delete Store
3- OpenAPI Aggregator watch OpenAPISpecStore and aggregate changes

sttts replied:
not sure what you mean with the store here. The basic pattern is that a controller uses a shared informer. This decouples the controllers while sharing a connection to the apiserver. So it's just a matter of plumbing your controller with its own shared informer. Should be pretty straight forward. Just look at some of the many examples.

In your control loop you can filter by the Available condition.

mbohlool replied:
I will look at some examples to understand this.

mbohlool replied:
> sharing a connection to the apiserver
That would apply to informers on rest endpoints/api objects.

Is there an example of custom informer? i.e. informer for a non-api object?

sttts replied:
They only exist for API resources.

mbohlool replied:
@deads2k I've discussed this in details with @sttts. Thought it looks easy to do at the beginning, it is quite complex to do. Right now OpenAPI controller gets its signal when Add/Remove APIService in APIAggregator is being called. That is all it needs to function. But there is quite a logic in APIServiceRegistrationController that listen on two informers, combine their signals with some logic and calls AddAPIService and RemoveAPIService. If we want to decouple OpenAPI controller from those two functions we end up either duplicating the code in APIServiceRegistrationController or factor them out somehow to be used in both controllers or have OpenAPIAggregationController derive from APIServiceRegistrationController and changes the queue and sync function (and log messages, etc.). All of these solutions are quite complex and we think it does not worth to do, unless we missed something here. would you agree that this decoupling is not crucial to this PR and current solutions are quite complex that deserve their own time to fix with a new PR (if we found a simple solution to it)?

sttts replied:
not to forget that we also need access to the `ProxyHandler`. So we get a couple of new (code) dependencies in exchange of the three add/remove/update callbacks. I would prefer the later.

deads2k replied:
> would you agree that this decoupling is not crucial to this PR and current solutions are quite complex that deserve their own time to fix with a new PR (if we found a simple solution to it)?

Thanks for looking at it.  At some point we should restructure in a way that makes it more reasonable, but I'm ok defering as long as this is non-blocking as it is.  My recollection is that is already done.

mbohlool replied:
Yes, it is non-blocking. Thanks.

deads2k commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go:
```diff
@@ -132,36 +166,46 @@ func sortByPriority(specs []openAPISpecInfo) {
 }
 
-// buildOpenAPISpec aggregates all OpenAPI specs.  It is not thread-safe.
+// buildOpenAPISpec aggregates all OpenAPI specs.  It is not thread-safe. The caller is responsible to hold proper locks.
```
then this should probably be `buildOpenAPISpec_locked`

mbohlool replied:
is using "_" in names recommended? `buildOpenAPISpec` is a private method anyway. it only used in this type.

sttts replied:
golint hates underscores

deads2k commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_controller.go:
```diff
@@ -0,0 +1,150 @@
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.Infof("Starting OpenAPIAggregationController")
```
I'd suggest waiting for your cache to prime

mbohlool replied:
which cache?

sttts replied:
If you have your own informer, you can (should) wait for it to be primed.

deads2k commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_controller.go:
```diff
@@ -0,0 +1,150 @@
+}
+
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (c *OpenAPIAggregationController) processNextWorkItem() bool {
```
This looks pretty close to a controller now.  This one should follow the normal conventions from the rest of our controllers and it will be better separated and easier for general maintainers to inspect/fix/contribute.

deads2k replied:
See https://github.com/kubernetes/community/blob/master/contributors/devel/controllers.md for the conventional form of processNextWorkItem.  Substituting just `syncHandler` makes the error handling easy to see as "standard" and keeps your logic out of the error handling paths.

mbohlool replied:
Done in next commit.

deads2k commented:
I like this new structure better, it's easier for me to reason about, but I think we should make this a full-fledged controller in its own right.  I don't think we want an inbetween phase where it's *almost*, but not quite an independent controller.

mbohlool commented:
> I like this new structure better, it's easier for me to reason about, but I think we should make this a full-fledged controller in its own right. I don't think we want an inbetween phase where it's almost, but not quite an independent controller.

What do you mean by full-fledged controller?

frobware commented:
@mbohlool adding the following test to `cmd/kube-apiserver/app/testing/server_test.go`:

    func TestTeardown(t *testing.T) {
        _, tearDown := StartTestServerOrDie(t)
        defer tearDown()
        <-make(chan struct{})
    }

I see merge errors if I let the test run and not exit:

    openapi_controller.go:107] aggregating OpenAPI spec for "v1beta1.authorization.k8s.io" failed with : unable to merge: duplicated path /apis/authentication.k8s.io/v1/tokenreviews


That's just an example, you can see more [here](https://gist.github.com/frobware/c07c12f4088f904e2ffef87d7336a4c8) - this runs for longer and you can the repeated merge attempts.

frobware commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_controller.go:
```diff
@@ -0,0 +1,150 @@
+
+func (c *OpenAPIAggregationController) AddAPIService(handler http.Handler, apiService *apiregistration.APIService) {
+	if err := c.openAPIAggregationManager.AddUpdateApiService(handler, apiService); err != nil {
+		utilruntime.HandleError(fmt.Errorf("adding %q to OpenAPIAggregationController failed with : %v", apiService.Name, err))
```
nit: drop space between between with and ':'. 

So, `with:`.

Ditto for the two other error messages in this file.

mbohlool replied:
Fixed.

mbohlool replied:
Done in next commit.

sttts commented:
@mbohlool any progress here? 1.8 code freeze is approaching.

deads2k commented:
/approve

@sttts it looked pretty good to me

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_controller.go:
```diff
@@ -0,0 +1,150 @@
+/*
```
if the dependencies allow it, moving this and possibly the aggregator to pkg/controllers/openapi would be great.

mbohlool replied:
Done in next commit.

mbohlool commented:
@frobware nice catch. There were a bug in kube-openapi. fixed here: https://github.com/kubernetes/kube-openapi/pull/8

Can you or @sttts review it please?

mbohlool commented:
waiting for kubernetes/kube-openapi#8 to be merged so I can vendor it in here and push new changes.

sttts commented on cmd/kube-apiserver/app/testing/server_test.go:
```diff
@@ -37,6 +37,13 @@ import (
 )
 
+
+func TestTeardown(t *testing.T) {
```
should this be here?

mbohlool replied:
ops. No. Will remove it.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/openapi_aggregator.go:
```diff
@@ -0,0 +1,314 @@
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
```
nit: you can get rid of the openapi_ prefixes in the file name now.

mbohlool replied:
Good idea.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/openapi_controller.go:
```diff
@@ -0,0 +1,167 @@
+// processNextWorkItem deals with one key off the queue.  It returns false when it's time to quit.
+func (c *OpenAPIAggregationController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	defer c.queue.Done(key)
```
seeing the Done behind the if-clause in all other controllers.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/openapi_controller.go:
```diff
@@ -0,0 +1,167 @@
+
+	err := c.syncHandler(key.(string))
+	if err == nil {
+		return true
```
Should we forget the key to reset rate limiting for that key? Most controllers do so, or have an explicit comment here if they don't.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/openapi_controller.go:
```diff
@@ -0,0 +1,167 @@
+func (c *OpenAPIAggregationController) sync(key string) error {
+	handler, etag, exists := c.openAPIAggregationManager.GetApiServiceInfo(key)
+	if !exists || handler == nil {
+		c.queue.Forget(key)
```
can't we move the Forget call into processNextWorkItem on `err == nil`?

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/openapi_controller.go:
```diff
@@ -0,0 +1,167 @@
+	}
+
+	if rateLimit {
+		c.queue.AddRateLimited(key)
```
This is basically your error case. Just return an error and let the `c.queue.AddRateLimited` do its work in processNextWorkItem.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/openapi_controller.go:
```diff
@@ -0,0 +1,167 @@
+		c.queue.AddRateLimited(key)
+	} else {
+		c.queue.Forget(key)
+		c.queue.AddAfter(key, successfulUpdateDelay)
```
What happens on a `AddAfter(key); Forget(key)`. Will the forget cancel the AddAfter?

sttts commented:
Looks good now :+1: Can you sqash? All generated code commits into one, the godeps one alone, the delegation chain alone and everything around openapi into one.

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go:
```diff
@@ -0,0 +1,169 @@
+}
+
+func (c *OpenAPIAggregationController) AddAPIService(handler http.Handler, apiService *apiregistration.APIService) {
+	if err := c.openAPIAggregationManager.AddUpdateApiService(handler, apiService); err != nil {
```
@mbohlool in @frobware's tests we see download errors for the local api groups after some minutes. Can it be that we have to filter out the local services from the controller logic?

sttts replied:
This plus squashing blocks my lgtm.

mbohlool replied:
This should be fixed with missing case for unmodified files. The squashing however is not that easy The delegation chain commit is in between OpenAPI commits. I had many commits and tried my best to reduce number of commits and make them as self contain and descriptive as possible. Let see how do you like that. Will push and ping you when it is ready.

sttts replied:
I was running @frobware's test and noticed what I suspected above: all local APIServices cause a download of the local spec. This takes roughly one second per APIService (i.e. per GroupVersion) on my machine, i.e. I see downloads and a busy CPU for rougly 20 seconds. After this the etag is recorded and the download is quick. But after each download the filtered spec is stored in the aggregator spec map. So you have each GVK spec twice in there: once from the delegation chain, once from the download. This slows down the merging by at least a factor of two (if don't have quadratic runtime somewhere, then it much worse). And if one of the spec changes over time, i.e. from CRDs (if we support them in OpenAPI), you have two name conflicts and renaming in the merge algorithm.

As before, I suggest to stop downloading local specs for now.

frobware commented:
Assuming the following test function in cmd/kube-apiserver/app/testing/server_test.go

```
func TestTeardown(t *testing.T) {
	config, tearDown := StartTestServerOrDie(t)
	defer tearDown()
	<-make(chan struct{})
}
```

I see messages like the following printed for different endpoints:

```
E0830 ... loading OpenAPI spec for "v1alpha1.rbac.authorization.k8s.io" failed with: OpenAPI spec does not exists
```

To reproduce I am doing:  `$ ./testing.test -test.run TestTeardown`, then letting that sit there, doing nothing else. I don't believe I have seen this before.

Here's more output:

https://gist.github.com/4e04128c301b1c07c13ef48871e61d7b

frobware commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go:
```diff
@@ -0,0 +1,169 @@
+	switch {
+	case err != nil:
+		return syncRequeueRateLimited, err
+	case httpStatus == http.StatusNotFound || returnSpec == nil:
```
There's no case that handles `http.StatusNotModified`.

I believe this is why after some period of time we get the error message(s):

```
controller.go:107] loading OpenAPI spec for "v2alpha1.batch" failed with: OpenAPI spec does not exists
```
Presumably it works OK first time but any subsequent downloads actually return a 304 (i.e., StatusNotModified).

mbohlool replied:
Yep, that case should not have been removed in the refactoring. I added it back with a no-op.

sttts replied:
`v2alpha1.batch` should not be downloaded in the first place as a local service.

mbohlool replied:
Those are local specs. They have handler (kubeApiServer) so they will be downloaded. But it is fast and one time. This can be improved for sure.

sttts replied:
but the local specs have there special `localDelegateChainNamePattern` key, don't they? Will they appear twice in the map then?

mbohlool committed:
Update Godep for kube-openapi

mbohlool committed:
update bazel

update OpenAPI spec

update staging godeps

mbohlool commented:
@sttts @frobware  fixed the issue with tests and squashed commits. PTAL.

mbohlool committed:
Provide whole delegate chain to kube aggregator

mbohlool commented:
/test pull-kubernetes-unit

mbohlool commented:
/test pull-kubernetes-e2e-gce-etcd3

mbohlool commented:
@smarterclayton @lavalamp This PR needs your approval. Thanks.

sttts commented:
https://github.com/kubernetes/kubernetes/pull/50864#discussion_r136320435 is still open, blocking the lgtm. But it is fine for approval.

@smarterclayton can you approve?

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go:
```diff
@@ -0,0 +1,328 @@
+	s.localAPIServiceNames = append(s.localAPIServiceNames, name)
+}
+
+// GetLocalDelegateNames returns list of delegate names. The controller use list list to update local delegates' spec.
```
this list

sttts commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/controller.go:
```diff
@@ -0,0 +1,185 @@
+		downloader: downloader,
+	}
+
+	for _, name := range openAPIAggregationManager.GetLocalDelegateNames() {
```
Why do we need this if all local specs are constant? Moreover, having the fake delegate chain names here in the controller seem to be wrong. They are no APIServices. They are the result of no better way of getting the openapi spec for local APIServices without iterating over the delegation chain. Now we leak this hack here.

mbohlool replied:
This is just an optimization vs using APIServices. We need to check local delegates for change in spec as CRDs will eventually show up there. The assumptions that local delegates are static is not always true while it is with current implementations.

sttts commented:
@mbohlool let me verify whether I understand correctly how local services are excluded from being downloaded:
- AddUpdateAPIService ignores local services by returning nil. The new service actually is not added. 
- The sync queue sees the local service and asks the aggregator via GetApiServiceInfo. Because it was never added, it returns nil and the sync queue stops considering the service.
- At the same time we leak the delegation chain fake names into the controller and download the specs using the preexisting ApiServiceInfos. We do this even though we know that local specs are constant and the download is not necessary. Why?

I am still not convinced that we on the right track. Considering the time before the freeze, I again suggest the following:
- stop dynamically updating local specs. This is feature is not ready. 
- add a check in the three entry points into the controller to ignore local specs completely from entering the queue. 
- reconsider a good structure when we actually get dynamic local specs for CRDs after 1.8. 

This is a 6 line change (after removing the leaked fake delagation chain service names) and we are good to go before the freeze. 

We do not loose anything with that and we keep the code simple. I had a hard time to understand the flow here in the PR with the latest local spec changes and when merged it’s even harder to follow.

sttts commented:
/lgtm

sttts commented:
/lgtm

sttts commented:
/retest

mbohlool commented:
/test pull-kubernetes-kubemark-e2e-gce

sttts commented:
/retest

sttts commented:
/retest

smarterclayton commented:
/approve

sttts commented:
/retest

sttts commented:
/retest

lavalamp commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go:
```diff
@@ -0,0 +1,318 @@
+
+type specAggregator struct {
+	// mutex protects all members of this struct.
+	rwMutex sync.RWMutex
```
"lock" is a common name for this. (not worth changing in this PR)

lavalamp commented on staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator.go:
```diff
@@ -0,0 +1,318 @@
+	localDelegateChainNamePattern = "k8s_internal_local_delegation_chain_%010d"
+
+	// A randomly generated UUID to differentiate local and remote eTags.
+	locallyGeneratedEtagPrefix = "\"6E8F849B434D4B98A569B9D7718876E9-"
```
This is a bit surprising, I would have expected this to be generated at runtime. I don't think there's too much mischief that can be done by knowing this, but it doesn't feel great. This can be fixed in a follow-up.

lavalamp commented:
/lgtm
/approve

mbohlool commented:
/test pull-kubernetes-kubemark-e2e-gce

sttts commented:
/lgtm

sttts commented:
/retest
